lib/versioner/util.go

```diff
@@ -38,11 +57,215 @@ func UntagFilename(path string) (string, string) {
 	versionTag := ExtractTag(path)
 
 	// Files tagged with old style tags cannot be untagged.
-	if versionTag == "" || strings.HasSuffix(ext, versionTag) {
+	if versionTag == "" {
 		return "", ""
 	}
 
+	// Old style tag
+	if strings.HasSuffix(ext, versionTag) {
+		return strings.TrimSuffix(path, "~"+versionTag), versionTag
+	}
+
 	withoutExt := path[:len(path)-len(ext)-len(versionTag)-1]
 	name := withoutExt + ext
 	return name, versionTag
 }
+
+func retrieveVersions(fileSystem fs.Filesystem) (map[string][]FileVersion, error) {
+	files := make(map[string][]FileVersion)
+
+	err := fileSystem.Walk(".", func(path string, f fs.FileInfo, err error) error {
+		// Skip root (which is ok to be a symlink)
+		if path == "." {
+			return nil
+		}
+
+		// Skip walking if we cannot walk...
+		if err != nil {
+			return err
+		}
+
+		// Ignore symlinks
+		if f.IsSymlink() {
+			return fs.SkipDir
+		}
+
+		// No records for directories
+		if f.IsDir() {
+			return nil
+		}
+
+		path = osutil.NormalizedFilename(path)
+
+		name, tag := UntagFilename(path)
+		// Something invalid, assume it's an untagged file
+		if name == "" || tag == "" {
+			versionTime := f.ModTime().Truncate(time.Second)
+			files[path] = append(files[path], FileVersion{
+				VersionTime: versionTime,
+				ModTime:     versionTime,
+				Size:        f.Size(),
+			})
+			return nil
+		}
+
+		versionTime, err := time.ParseInLocation(TimeFormat, tag, locationLocal)
+		if err != nil {
+			// Can't parse it, welp, continue
+			return nil
+		}
+
+		if err == nil {
+			files[name] = append(files[name], FileVersion{
+				VersionTime: versionTime.Truncate(time.Second),
+				ModTime:     f.ModTime().Truncate(time.Second),
+				Size:        f.Size(),
+			})
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return files, nil
+}
+
+type fileTagger func(string, string) string
+
+func archiveFile(srcFs, dstFs fs.Filesystem, filePath string, tagger fileTagger) error {
+	filePath = osutil.NativeFilename(filePath)
+	info, err := srcFs.Lstat(filePath)
+	if fs.IsNotExist(err) {
+		l.Debugln("not archiving nonexistent file", filePath)
+		return nil
+	} else if err != nil {
+		return err
+	}
+	if info.IsSymlink() {
+		panic("bug: attempting to version a symlink")
+	}
+
+	_, err = dstFs.Stat(".")
+	if err != nil {
+		if fs.IsNotExist(err) {
+			l.Debugln("creating versions dir")
+			err := dstFs.Mkdir(".", 0755)
+			if err != nil {
+				return err
+			}
+			_ = dstFs.Hide(".")
+		} else {
+			return err
+		}
+	}
+
+	l.Debugln("archiving", filePath)
+
+	file := filepath.Base(filePath)
+	inFolderPath := filepath.Dir(filePath)
+
+	err = dstFs.MkdirAll(inFolderPath, 0755)
+	if err != nil && !fs.IsExist(err) {
+		return err
+	}
+
+	ver := tagger(file, info.ModTime().Format(TimeFormat))
```

Comment:
Followup to comment on https://github.com/syncthing/syncthing/pull/5514/files#diff-607726c1924296ce5c4bb168265bf986R138 :
Replace with something like 
```
ver := file
if !noTag { ver = TagFilename(...) }
```